### PR TITLE
Implement labelling for release notes and topics check

### LINKED ISSUE #68459: Soft mandatory labeling for release notes
Composing release notes has historically been a very time consuming process. Until now, we would have one dedicated engineer from the PyTorch Core team manually categorize all the commits (usually about 3-3.5k commits)  every release, which is usually about a week's worth of full-time work. 

In PyTorch 1.10, we made a significant improvement, by auto-categorizing about 30% (1k) of the 3.5k commits using [this](https://github.com/pytorch/pytorch/pull/64929) script that labeled PRs based on the files changed. 

Now, we are proposing to move to an upfront labeling system, where all PR authors categorize and topic each PR upfront, which will make compiling release notes much easier at the time of release with very little upfront work. This would mean that in the future releases, categorizing of PRs for compiling release notes will be solely done based on the category labels. 

Inspired by https://github.com/pytorch/vision/pull/4618 we propose to add GitHub workflow/Probot path that would remind authors to assign *one* of the following category labels to PRs:

- `release notes: mobile`
- `release notes: jit`
- `release notes: visualization`
- `release notes: onnx`
- `release notes: quantization`
- `release notes: rocm`
- `release notes: cuda`
- `release notes: benchmark`
- `release notes: profiler`
- `release notes: performance_as_product`
- `release notes: package/deploy`
- `release notes: composability`
- `release notes: releng`
- `release notes: fx`
- `release notes: vulkan`
- `release notes: nn`
- `release notes: cpp`
- `release notes notes: python_frontend`
- `release notes: complex`
- `release notes: autograd`
- `release notes: build`
- `release notes: memory format`
- `release notes: foreach_frontend`
- `release notes: dataloader`
- `release notes: linalg_frontend`
- `release notes: Meta API`
- `release notes: sparse`
- `release notes: distributed (c10d)`
- `release notes: distributed (rpc)`
- `release notes: distributed (ddp)`
- `release notes: distributed (pipeline)`
- `release notes: distributed (fsdp)`
- `release notes: distributed (sharded)`

Please leave a comment, with additional category labels if you are working on a new feature or want to refine an existing category label into more specific category labels to distribute the work or for other reasons. Example., at the time of writing this, `distributed` category is split into six categories (`c10d`, `rpc`, `ddp`, `pipeline`, `fsdp`, `sharded`)

And optionally, assign *one* of the following topic labels:

- `topic: bc_breaking`
- `topic: deprecation`
- `topic: new feature`
- `topic: improvements`
- `topic: bug fixes`
- `topic: performance`
- `topic: documentation`
- `topic: developer feature`
- `topic: non-user visible` <- assign this label if you think PR should not be mentioned in release notes

Later assigning such labels might become a mandatory pre-merge check

cc @brianjo @mruberry @seemethere @malfet @pytorch/pytorch-dev-infra

### PR DESCRIPTION
**Pull request by atalman**
Fixes #68459

### CODE REVIEW

File: .github/scripts/process_commit.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
+"""
```
Comment (malfet) [inline .github/scripts/process_commit.py:1]: Please add shebang
```suggestion
#!/usr/bin/env python3
"""
```

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
Also, it would be nice to mention that script is inspired by https://github.com/pytorch/vision/blob/main/.github/process_commit.py

File: .github/scripts/process_commit.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+import sys
+from typing import Any, Optional, Set, Tuple
+
+import requests
+
+# For a PR to be properly labeled it should have one primary label and one secondary label
+
+RELEASE_NOTES_LABELS = {
```
Comment (anjali411) [inline .github/scripts/process_commit.py:19]: alternately is there a to pull out all the `release notes: ...` labels here instead of manually mentioning a list. This would make it easier for people to add more release notes labels just via github (without a codechange) if they want to.

Comment (atalman) [reply .github/scripts/process_commit.py:19]: Great idea! Let me look into it!

**Comment by anjali411** [comment]
Also, before we merge this PR we should actually add all the release notes category and topic labels. 
As a way to be transparent and avoid confusion, should we have the `facebook-github-bot` link the issue containing available release categories and labels in its comment on each PR?

File: .github/scripts/process_commit.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+"""
+
+import sys
+from typing import Any, Set, Tuple
+
+import requests
+
+# For a PR to be properly labeled it should have release notes label and one topic label
+PRIMARY_LEBEL_FILTER = "release notes:"
+SECONDARY_LEBEL_FILTER = "topic:"
```
Comment (anjali411) [inline .github/scripts/process_commit.py:19]: I think it actually might be better to explicitly list the topics here since we don't generally expect the topics to change (they are consistent across all categories).

File: .github/scripts/process_commit.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    if not pr_number:
+        sys.exit(0)
+
+    merger, labels = get_pr_merger_and_labels(pr_number)
+    response = query_pytroch("labels", accept="application/json")
+    response_labels = list(map(lambda x: str(x["name"]), response.json()))
+    primary_labels = set(filter(lambda x: x.startswith(PRIMARY_LEBEL_FILTER), response_labels))
+    secondary_labels = set(filter(lambda x: x.startswith(SECONDARY_LEBEL_FILTER), response_labels))
+
+    is_properly_labeled = bool(primary_labels.intersection(labels) and secondary_labels.intersection(labels))
```
Comment (anjali411) [inline .github/scripts/process_commit.py:55]: @malfet do we wanna mandate people to add labels upfront too?

File: .github/workflows/pr-labels.yml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      - name: Ping merger responsible for labeling if necessary
+        if: ${{ steps.commit.outputs.merger != '' }}
+        uses: mshick/add-pr-comment@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          message: |
+            Hey ${{ steps.commit.outputs.merger }}!
+
+            You merged this PR, but no labels were added. The list of valid labels is available at https://github.com/pytorch/pytorch/blob/master/.github/scripts/process_commit.py
```
Comment (anjali411) [inline .github/workflows/pr-labels.yml:39]: nit - but no release notes category and topic labels were added

File: .github/scripts/process_commit.py (hunk 3 above)
Comment (atalman) [reply .github/scripts/process_commit.py:19]: @anjali411 for consistency I would rather use same approach for release notes and topics. As you mentioned before, its easier to curate the release notes items and topics simply by adding them to available labes.
However with your new suggestion, we want "release notes" to be retrieved from github available labels and topics hardcoded ?

File: .github/scripts/process_commit.py (hunk 4 above)
Comment (atalman) [reply .github/scripts/process_commit.py:55]: > Also, before we merge this PR we should actually add all the release notes category and topic labels. As a way to be transparent and avoid confusion, should we have the `facebook-github-bot` link the issue containing available release categories and labels in its comment on each PR?

@anjali411 and @malfet Acknoweldge before merging, will add necessary labels

File: .github/scripts/process_commit.py (hunk 3 above)
Comment (anjali411) [reply .github/scripts/process_commit.py:19]: yeah I think it might be better to hardcode topics because we don't usually want to add new topics and we shouldn't allow people to add new topics without a consensus

File: .github/scripts/process_commit.py (hunk 4 above)
Comment (atalman) [reply .github/scripts/process_commit.py:55]: > @malfet do we wanna mandate people to add labels upfront too?
@anjali411 Can you please explain the definition of mandate, prevent from merging ? 

What is currently implemented is simply pinging people, if the labels are not applied, perhaps we should implement this as a first stage. And enforcing this policy can be a followup PR. I guess we will have to notify people that there is this new policy in place etc...

File: .github/scripts/process_commit.py (hunk 3 above)
Comment (atalman) [reply .github/scripts/process_commit.py:19]: Ok Sounds good. Let me do it the way you suggested.

File: .github/scripts/process_commit.py (hunk 4 above)
Comment (anjali411) [reply .github/scripts/process_commit.py]: sorry I meant add topic labels upfront

**Review by anjali411 — APPROVED** [review]
LGTM thank you!
